server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/InfotonReporter.scala

```diff
@@ -131,38 +128,37 @@ class InfotonReporter private (baseUrl: String, path: String)(implicit mat: Mate
       .map(_.payload)
   }
 
-  override def saveTokens(tokenAndStatistics: TokenAndStatisticsMap): Unit = {
-    def createRequest(tokensStats: TokenAndStatisticsMap) = {
-      val data = HttpEntity(
-        tokensStats
-          .foldLeft(Seq.empty[String]) {
-            case (agg, (sensor, (token, downloadStats))) => agg :+ createTriples(sensor, token, downloadStats)
-          }
-          .mkString("\n")
-      )
-      HttpRequest(uri = s"http://$host:$port/_in?format=$format&replace-mode", method = HttpMethods.POST, entity = data)
-        .addHeader(RawHeader("X-CM-WELL-TOKEN", writeToken))
+  def saveTokens(tokenAndStatistics: TokenAndStatisticsMap) : Unit = {
+
+    def createJson(sensor: String, token: Token, downloadStats: Option[DownloadStats]) = {
+      Json.fromFields({
+        downloadStats match {
+          case None => List(("sensor", Json.fromString(sensor)), ("token", Json.fromString(token)))
+          case Some(downloadStats) =>
+            List(("sensor", Json.fromString(sensor)), ("token", Json.fromString(token)),
+              ("receivedInfotons", Json.fromLong(downloadStats.receivedInfotons)))
+        }
+      }).noSpaces
     }
 
-    def createTriples(sensor: String, token: Token, downloadStats: Option[DownloadStats]) = {
-      val p = if (path.startsWith("/")) path.tail else path
-      s"""<cmwell://$p/tokens/$sensor> <cmwell://meta/nn#token> "$token" .""" + downloadStats.fold("")({ stats =>
-        "\n" + s"""<cmwell://$p/tokens/$sensor> <cmwell://meta/nn#receivedInfotons> "${stats.receivedInfotons}" ."""
-      })
+    def createTokenPayload(tokenAndStatistics: TokenAndStatisticsMap)  = {
+      tokenAndStatistics.foldLeft(Seq.empty[String]) {
+        case (agg, (sensor, (token, downloadStats))) => agg :+ createJson(sensor, token, downloadStats)
+      }.mkString("\n")
     }
 
     Source
       .single(tokenAndStatistics)
-      .map(createRequest)
-      .via(Http(context.system).outgoingConnection(host, port))
-      .map {
-        case HttpResponse(s, h, e, _) if s.isSuccess() =>
-          logger.debug(s"successfully written tokens infoton to $path")
-          e.discardBytes()
-        case HttpResponse(s, h, e, _) =>
-          logger.error(s"problem writing tokens infoton to $path")
-          e.discardBytes()
+      .map{createTokenPayload}
+      .map{
+        zStore.putString(s"stp-agent-${extractLastPart(path)}", _)
       }
+      .map(zStoreFuture => {
+        zStoreFuture.onComplete {
+          case Success(_) => logger.debug(s"successfully written tokens to zStore, key=$path")
+          case Failure(ex) => logger.error(s"problem writing tokens to zStore: ${ex.getMessage}")
```

Comment:
it is better to `logger.error("explanation...", ex)` - in that case you have all StackTrace.